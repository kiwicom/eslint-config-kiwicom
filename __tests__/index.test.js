// Finds rules that are supported by ESLint but not defined in our config.
// Because explicit is better than implicit.

// @flow

const path = require('path');
const eslintRules = require('eslint/lib/load-rules');

// $FlowAllowDynamicImport
const eslintConfig = require(path.join(__dirname, '..', 'index.js'));

const ourRules = new Set(Object.keys(eslintConfig.rules));
const supportedRules = new Set(Object.keys(eslintRules()));

// see: https://eslint.org/docs/rules/#deprecated
const deprecatedRules = new Set([
  'indent-legacy', // indent
  'lines-around-directive', // padding-line-between-statements
  'newline-after-var', // padding-line-between-statements
  'newline-before-return', // padding-line-between-statements
  'no-catch-shadow', // no-shadow
  'no-native-reassign', // no-global-assign
  'no-negated-in-lhs', // no-unsafe-negation
  'no-spaced-func', // func-call-spacing
  'prefer-reflect', // (no replacement)

  // https://github.com/babel/eslint-plugin-babel#deprecated
  'babel/array-bracket-spacing', // array-bracket-spacing
  'babel/arrow-parens', // arrow-parens
  'babel/flow-object-type', // flowtype/object-type-delimiter
  'babel/func-params-comma-dangle', // comma-dangle
  'babel/generator-star-spacing', // generator-star-spacing
  'babel/no-await-in-loop', // no-await-in-loop
  'babel/object-shorthand', // object-shorthand
]);

// Get plugins from package.json. Assume they're all in dependencies.
const plugins = Object.keys(require('../package.json').dependencies)
  .filter(dep => dep.startsWith('eslint-plugin'))
  .filter(dep => {
    // we are not testing every 3rd party plugins yet
    const whitelistedPlugins = ['eslint-plugin-babel'];
    return whitelistedPlugins.includes(dep);
  })
  .map(dep => dep.replace('eslint-plugin-', ''));

plugins.forEach(plugin => {
  // $FlowAllowDynamicImport
  Object.keys(require(`eslint-plugin-${plugin}`).rules).forEach(rule => {
    supportedRules.add(`${plugin}/${rule}`);
  });
});

const missing = new Set();
const extra = new Set();
const deprecated = new Set();

ourRules.forEach(rule => {
  if (deprecatedRules.has(rule)) {
    deprecated.add(rule);
  }
});

deprecatedRules.forEach(rule => {
  ourRules.add(rule);
});

ourRules.forEach(rule => {
  if (!supportedRules.has(rule)) {
    extra.add(rule);
  }
});

supportedRules.forEach(rule => {
  if (!ourRules.has(rule)) {
    missing.add(rule);
  }
});

test('missing Eslint rules', () => {
  expect(missing).toEqual(new Set());
});

test('deprecated Eslint rules', () => {
  expect(deprecated).toEqual(new Set());
});

// TODO: test extra rules:
// console.log('extra', extra); // eslint-disable-line no-console
