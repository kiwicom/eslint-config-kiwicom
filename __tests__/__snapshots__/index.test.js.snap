// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rules snapshot 1`] = `
Object {
  "globals": Object {
    "FormData": false,
    "__": false,
    "__DEV__": false,
  },
  "plugins": Array [
    "eslint-plugin-babel",
    "eslint-plugin-flowtype",
    "eslint-plugin-jest",
    "eslint-plugin-prefer-object-spread",
    "eslint-plugin-react",
    "eslint-plugin-react-hooks",
    "eslint-plugin-jsx-a11y",
    "eslint-plugin-react-native",
    "eslint-plugin-relay",
    "eslint-plugin-import",
    "eslint-plugin-dependencies",
    "eslint-plugin-monorepo",
    "eslint-plugin-node",
    "eslint-plugin-eslint-comments",
    "eslint-plugin-kiwicom-incubator",
    "eslint-plugin-prettier",
  ],
  "rules": Object {
    "accessor-pairs": Array [
      1,
      Object {
        "setWithoutGet": true,
      },
    ],
    "array-bracket-newline": "off",
    "array-bracket-spacing": "off",
    "array-callback-return": 0,
    "array-element-newline": "off",
    "arrow-body-style": 0,
    "arrow-parens": "off",
    "arrow-spacing": "off",
    "babel/camelcase": Array [
      1,
      Object {
        "ignoreDestructuring": false,
        "properties": "never",
      },
    ],
    "babel/new-cap": 0,
    "babel/no-invalid-this": 0,
    "babel/no-unused-expressions": 2,
    "babel/object-curly-spacing": "off",
    "babel/quotes": 0,
    "babel/semi": "off",
    "babel/valid-typeof": 2,
    "block-scoped-var": 0,
    "block-spacing": "off",
    "brace-style": "off",
    "callback-return": 0,
    "camelcase": 0,
    "capitalized-comments": 0,
    "class-methods-use-this": 0,
    "comma-dangle": "off",
    "comma-spacing": "off",
    "comma-style": "off",
    "complexity": 0,
    "computed-property-spacing": "off",
    "consistent-return": Array [
      1,
      Object {
        "treatUndefinedAsUnspecified": false,
      },
    ],
    "consistent-this": 0,
    "constructor-super": 2,
    "curly": 0,
    "default-case": 0,
    "dependencies/case-sensitive": 2,
    "dependencies/no-cycles": 1,
    "dependencies/no-unresolved": 0,
    "dependencies/require-json-ext": 2,
    "dot-location": "off",
    "dot-notation": 1,
    "eol-last": "off",
    "eqeqeq": Array [
      1,
      "smart",
    ],
    "eslint-comments/disable-enable-pair": 0,
    "eslint-comments/no-aggregating-enable": 0,
    "eslint-comments/no-duplicate-disable": 1,
    "eslint-comments/no-restricted-disable": 0,
    "eslint-comments/no-unlimited-disable": 0,
    "eslint-comments/no-unused-disable": 0,
    "eslint-comments/no-unused-enable": 1,
    "eslint-comments/no-use": 0,
    "flowtype/array-style-complex-type": 0,
    "flowtype/array-style-simple-type": 0,
    "flowtype/boolean-style": "off",
    "flowtype/define-flow-type": 1,
    "flowtype/delimiter-dangle": "off",
    "flowtype/generic-spacing": "off",
    "flowtype/newline-after-flow-annotation": Array [
      1,
      "always",
    ],
    "flowtype/no-dupe-keys": 1,
    "flowtype/no-existential-type": 1,
    "flowtype/no-flow-fix-me-comments": 0,
    "flowtype/no-mixed": 0,
    "flowtype/no-mutable-array": 0,
    "flowtype/no-primitive-constructor-types": 1,
    "flowtype/no-types-missing-file-annotation": 1,
    "flowtype/no-unused-expressions": 1,
    "flowtype/no-weak-types": 0,
    "flowtype/object-type-delimiter": "off",
    "flowtype/require-compound-type-alias": 0,
    "flowtype/require-exact-type": 0,
    "flowtype/require-parameter-type": 0,
    "flowtype/require-return-type": 0,
    "flowtype/require-types-at-top": 0,
    "flowtype/require-valid-file-annotation": Array [
      1,
      "always",
    ],
    "flowtype/require-variable-type": 0,
    "flowtype/semi": "off",
    "flowtype/sort-keys": 0,
    "flowtype/space-after-type-colon": "off",
    "flowtype/space-before-generic-bracket": "off",
    "flowtype/space-before-type-colon": "off",
    "flowtype/spread-exact-type": 0,
    "flowtype/type-id-match": 0,
    "flowtype/type-import-style": 0,
    "flowtype/union-intersection-spacing": "off",
    "flowtype/use-flow-type": 1,
    "flowtype/valid-syntax": 0,
    "for-direction": 2,
    "func-call-spacing": "off",
    "func-name-matching": 0,
    "func-names": 0,
    "func-style": 0,
    "function-paren-newline": "off",
    "generator-star": "off",
    "generator-star-spacing": "off",
    "getter-return": 2,
    "global-require": 0,
    "guard-for-in": 0,
    "handle-callback-err": 1,
    "id-blacklist": 0,
    "id-length": 0,
    "id-match": 0,
    "implicit-arrow-linebreak": "off",
    "import/newline-after-import": 2,
    "import/no-absolute-path": 2,
    "import/no-anonymous-default-export": Array [
      1,
      Object {
        "allowAnonymousClass": false,
        "allowAnonymousFunction": false,
        "allowArray": true,
        "allowArrowFunction": false,
        "allowCallExpression": true,
        "allowLiteral": true,
        "allowObject": true,
      },
    ],
    "import/no-duplicates": 1,
    "import/no-dynamic-require": 0,
    "import/no-extraneous-dependencies": Array [
      1,
      Object {
        "devDependencies": Array [
          "**/*.test.js",
          "**/*.spec.js",
        ],
      },
    ],
    "import/no-mutable-exports": 2,
    "import/no-unresolved": Array [
      1,
      Object {
        "ignore": Array [
          "^__generated__/",
        ],
      },
    ],
    "import/no-useless-path-segments": 2,
    "import/order": Array [
      1,
      Object {
        "groups": Array [
          Array [
            "builtin",
            "external",
          ],
          Array [
            "parent",
            "sibling",
          ],
          "index",
        ],
        "newlines-between": "always",
      },
    ],
    "import/prefer-default-export": 0,
    "indent": "off",
    "indent-legacy": "off",
    "init-declarations": 0,
    "jest/consistent-test-it": 0,
    "jest/expect-expect": 0,
    "jest/lowercase-name": 0,
    "jest/no-alias-methods": 0,
    "jest/no-disabled-tests": 1,
    "jest/no-empty-title": 1,
    "jest/no-focused-tests": 2,
    "jest/no-hooks": 0,
    "jest/no-identical-title": 2,
    "jest/no-jasmine-globals": 2,
    "jest/no-jest-import": 2,
    "jest/no-large-snapshots": 0,
    "jest/no-mocks-import": 1,
    "jest/no-test-callback": 0,
    "jest/no-test-prefixes": 0,
    "jest/no-test-return-statement": 2,
    "jest/no-truthy-falsy": 0,
    "jest/prefer-called-with": 1,
    "jest/prefer-expect-assertions": 0,
    "jest/prefer-inline-snapshots": 0,
    "jest/prefer-spy-on": 0,
    "jest/prefer-strict-equal": 0,
    "jest/prefer-to-be-null": 2,
    "jest/prefer-to-be-undefined": 2,
    "jest/prefer-to-contain": 2,
    "jest/prefer-to-have-length": 2,
    "jest/prefer-todo": 2,
    "jest/require-tothrow-message": 0,
    "jest/valid-describe": 2,
    "jest/valid-expect": 1,
    "jest/valid-expect-in-promise": 2,
    "jsx-a11y/accessible-emoji": 2,
    "jsx-a11y/alt-text": 1,
    "jsx-a11y/aria-props": 2,
    "jsx-a11y/aria-unsupported-elements": 2,
    "jsx-a11y/no-distracting-elements": Array [
      1,
      Object {
        "elements": Array [
          "marquee",
          "blink",
        ],
      },
    ],
    "jsx-quotes": "off",
    "key-spacing": "off",
    "keyword-spacing": "off",
    "kiwicom-incubator/no-invalid-flow-annotations": 2,
    "kiwicom-incubator/only-nullable-fields": 2,
    "line-comment-position": 0,
    "linebreak-style": "off",
    "lines-around-comment": 0,
    "lines-between-class-members": Array [
      1,
      "always",
      Object {
        "exceptAfterSingleLine": true,
      },
    ],
    "max-classes-per-file": 0,
    "max-depth": 0,
    "max-len": 0,
    "max-lines": 0,
    "max-lines-per-function": 0,
    "max-nested-callbacks": 0,
    "max-params": 0,
    "max-statements": 0,
    "max-statements-per-line": 0,
    "monorepo/no-internal-import": 2,
    "monorepo/no-relative-import": 2,
    "multiline-comment-style": 0,
    "multiline-ternary": "off",
    "new-cap": Array [
      1,
      Object {
        "capIsNew": false,
        "newIsCap": true,
      },
    ],
    "new-parens": "off",
    "newline-per-chained-call": "off",
    "no-alert": 1,
    "no-array-constructor": 1,
    "no-arrow-condition": "off",
    "no-async-promise-executor": 1,
    "no-await-in-loop": 1,
    "no-bitwise": 1,
    "no-buffer-constructor": 1,
    "no-caller": 2,
    "no-case-declarations": 1,
    "no-class-assign": 1,
    "no-comma-dangle": "off",
    "no-compare-neg-zero": 2,
    "no-cond-assign": 1,
    "no-confusing-arrow": 0,
    "no-console": 2,
    "no-const-assign": 2,
    "no-constant-condition": Array [
      1,
      Object {
        "checkLoops": false,
      },
    ],
    "no-continue": 0,
    "no-control-regex": 1,
    "no-debugger": 2,
    "no-delete-var": 2,
    "no-div-regex": 1,
    "no-dupe-args": 2,
    "no-dupe-class-members": 2,
    "no-dupe-keys": 2,
    "no-duplicate-case": 1,
    "no-duplicate-imports": 0,
    "no-else-return": 2,
    "no-empty": Array [
      1,
      Object {
        "allowEmptyCatch": true,
      },
    ],
    "no-empty-character-class": 1,
    "no-empty-function": 0,
    "no-empty-pattern": 1,
    "no-eq-null": 0,
    "no-eval": 2,
    "no-ex-assign": 1,
    "no-extend-native": 1,
    "no-extra-bind": 1,
    "no-extra-boolean-cast": 1,
    "no-extra-label": 1,
    "no-extra-parens": "off",
    "no-extra-semi": "off",
    "no-fallthrough": 1,
    "no-floating-decimal": "off",
    "no-func-assign": 2,
    "no-global-assign": Array [
      1,
      Object {
        "exceptions": Array [
          "Map",
          "Set",
        ],
      },
    ],
    "no-implicit-coercion": 0,
    "no-implicit-globals": 0,
    "no-implied-eval": 2,
    "no-inline-comments": 0,
    "no-inner-declarations": 1,
    "no-invalid-regexp": 1,
    "no-invalid-this": 0,
    "no-irregular-whitespace": 1,
    "no-iterator": 1,
    "no-label-var": 2,
    "no-labels": Array [
      1,
      Object {
        "allowLoop": true,
        "allowSwitch": true,
      },
    ],
    "no-lone-blocks": 1,
    "no-lonely-if": 2,
    "no-loop-func": 0,
    "no-magic-numbers": 0,
    "no-misleading-character-class": 1,
    "no-mixed-operators": 0,
    "no-mixed-requires": 0,
    "no-mixed-spaces-and-tabs": "off",
    "no-multi-assign": 0,
    "no-multi-spaces": "off",
    "no-multi-str": 2,
    "no-multiple-empty-lines": "off",
    "no-negated-condition": 0,
    "no-nested-ternary": 0,
    "no-new": 1,
    "no-new-func": 2,
    "no-new-object": 1,
    "no-new-require": 1,
    "no-new-symbol": 1,
    "no-new-wrappers": 1,
    "no-obj-calls": 2,
    "no-octal": 1,
    "no-octal-escape": 1,
    "no-param-reassign": 2,
    "no-path-concat": 1,
    "no-plusplus": 0,
    "no-process-env": 0,
    "no-process-exit": 0,
    "no-proto": 2,
    "no-prototype-builtins": 0,
    "no-redeclare": Array [
      1,
      Object {
        "builtinGlobals": true,
      },
    ],
    "no-regex-spaces": 1,
    "no-reserved-keys": "off",
    "no-restricted-globals": 0,
    "no-restricted-imports": 0,
    "no-restricted-modules": 0,
    "no-restricted-properties": 0,
    "no-restricted-syntax": 0,
    "no-return-assign": 1,
    "no-return-await": 1,
    "no-script-url": 2,
    "no-self-assign": 1,
    "no-self-compare": 1,
    "no-sequences": 1,
    "no-shadow": 0,
    "no-shadow-restricted-names": 1,
    "no-space-before-semi": "off",
    "no-spaced-func": "off",
    "no-sparse-arrays": 2,
    "no-sync": 0,
    "no-tabs": 0,
    "no-template-curly-in-string": 0,
    "no-ternary": 0,
    "no-this-before-super": 2,
    "no-throw-literal": 1,
    "no-trailing-spaces": "off",
    "no-undef": 2,
    "no-undef-init": 0,
    "no-undefined": 0,
    "no-underscore-dangle": 0,
    "no-unexpected-multiline": 0,
    "no-unmodified-loop-condition": 0,
    "no-unneeded-ternary": 1,
    "no-unreachable": 2,
    "no-unsafe-finally": 1,
    "no-unsafe-negation": 2,
    "no-unused-expressions": 0,
    "no-unused-labels": 1,
    "no-unused-vars": Array [
      1,
      Object {
        "args": "after-used",
      },
    ],
    "no-use-before-define": 0,
    "no-useless-call": 1,
    "no-useless-catch": 1,
    "no-useless-computed-key": 1,
    "no-useless-concat": 1,
    "no-useless-constructor": 1,
    "no-useless-escape": 1,
    "no-useless-rename": 1,
    "no-useless-return": 1,
    "no-var": 1,
    "no-void": 1,
    "no-warning-comments": 0,
    "no-whitespace-before-property": "off",
    "no-with": 1,
    "no-wrap-func": "off",
    "node/exports-style": 0,
    "node/no-deprecated-api": 1,
    "node/no-extraneous-import": 0,
    "node/no-extraneous-require": 0,
    "node/no-missing-import": 0,
    "node/no-missing-require": 2,
    "node/no-unpublished-bin": 1,
    "node/no-unpublished-import": 0,
    "node/no-unpublished-require": 0,
    "node/no-unsupported-features/es-builtins": 0,
    "node/no-unsupported-features/es-syntax": 0,
    "node/no-unsupported-features/node-builtins": 0,
    "node/prefer-global/buffer": 1,
    "node/prefer-global/console": 1,
    "node/prefer-global/process": 1,
    "node/prefer-global/text-decoder": 0,
    "node/prefer-global/text-encoder": 0,
    "node/prefer-global/url": 1,
    "node/prefer-global/url-search-params": 0,
    "node/process-exit-as-throw": 0,
    "node/shebang": 1,
    "nonblock-statement-body-position": "off",
    "object-curly-newline": "off",
    "object-curly-spacing": "off",
    "object-property-newline": "off",
    "object-shorthand": 0,
    "one-var": Array [
      1,
      Object {
        "initialized": "never",
      },
    ],
    "one-var-declaration-per-line": "off",
    "operator-assignment": 1,
    "operator-linebreak": "off",
    "padded-blocks": "off",
    "padding-line-between-statements": 0,
    "prefer-arrow-callback": 0,
    "prefer-const": Array [
      1,
      Object {
        "destructuring": "all",
      },
    ],
    "prefer-destructuring": 0,
    "prefer-named-capture-group": 1,
    "prefer-numeric-literals": 0,
    "prefer-object-spread": 0,
    "prefer-object-spread/prefer-object-spread": 1,
    "prefer-promise-reject-errors": 1,
    "prefer-rest-params": 1,
    "prefer-spread": 1,
    "prefer-template": 0,
    "prettier/prettier": Array [
      2,
      Object {
        "bracketSpacing": true,
        "printWidth": 80,
        "singleQuote": true,
        "tabWidth": 2,
        "trailingComma": "all",
      },
    ],
    "quote-props": "off",
    "quotes": 0,
    "radix": 2,
    "react-hooks/exhaustive-deps": 1,
    "react-hooks/rules-of-hooks": 2,
    "react-native/no-unused-styles": 2,
    "react/display-name": 0,
    "react/forbid-component-props": 0,
    "react/forbid-prop-types": 0,
    "react/jsx-boolean-value": 0,
    "react/jsx-child-element-spacing": "off",
    "react/jsx-closing-bracket-location": "off",
    "react/jsx-closing-tag-location": "off",
    "react/jsx-curly-spacing": "off",
    "react/jsx-equals-spacing": "off",
    "react/jsx-filename-extension": 0,
    "react/jsx-first-prop-new-line": "off",
    "react/jsx-handler-names": 0,
    "react/jsx-indent": "off",
    "react/jsx-indent-props": "off",
    "react/jsx-key": 1,
    "react/jsx-max-props-per-line": "off",
    "react/jsx-no-bind": 2,
    "react/jsx-no-comment-textnodes": 1,
    "react/jsx-no-duplicate-props": 2,
    "react/jsx-no-literals": 0,
    "react/jsx-no-target-blank": 0,
    "react/jsx-no-undef": 2,
    "react/jsx-one-expression-per-line": "off",
    "react/jsx-pascal-case": 0,
    "react/jsx-props-no-multi-spaces": "off",
    "react/jsx-sort-props": 0,
    "react/jsx-space-before-closing": "off",
    "react/jsx-tag-spacing": "off",
    "react/jsx-uses-react": 1,
    "react/jsx-uses-vars": 1,
    "react/jsx-wrap-multilines": "off",
    "react/no-access-state-in-setstate": 2,
    "react/no-children-prop": 0,
    "react/no-danger": 0,
    "react/no-danger-with-children": 0,
    "react/no-deprecated": 1,
    "react/no-did-mount-set-state": 0,
    "react/no-did-update-set-state": 2,
    "react/no-direct-mutation-state": 1,
    "react/no-find-dom-node": 0,
    "react/no-is-mounted": 1,
    "react/no-multi-comp": 0,
    "react/no-render-return-value": 0,
    "react/no-set-state": 0,
    "react/no-string-refs": 0,
    "react/no-unescaped-entities": 0,
    "react/no-unknown-property": 1,
    "react/no-unused-prop-types": 0,
    "react/prefer-es6-class": 0,
    "react/prefer-stateless-function": 0,
    "react/prop-types": 0,
    "react/react-in-jsx-scope": 1,
    "react/require-optimization": 0,
    "react/require-render-return": 0,
    "react/self-closing-comp": "off",
    "react/sort-comp": 2,
    "react/sort-prop-types": 0,
    "react/style-prop-object": 0,
    "relay/compat-uses-vars": 0,
    "relay/generated-flow-types": 2,
    "relay/graphql-naming": 2,
    "relay/graphql-syntax": 2,
    "relay/no-future-added-value": 2,
    "relay/unused-fields": 2,
    "require-atomic-updates": 2,
    "require-await": 1,
    "require-jsdoc": 0,
    "require-unicode-regexp": 0,
    "require-yield": 0,
    "rest-spread-spacing": "off",
    "semi": "off",
    "semi-spacing": "off",
    "semi-style": "off",
    "sort-imports": 0,
    "sort-keys": 0,
    "sort-vars": 0,
    "space-after-function-name": "off",
    "space-after-keywords": "off",
    "space-before-blocks": "off",
    "space-before-function-paren": "off",
    "space-before-function-parentheses": "off",
    "space-before-keywords": "off",
    "space-in-brackets": "off",
    "space-in-parens": "off",
    "space-infix-ops": "off",
    "space-return-throw-case": "off",
    "space-unary-ops": "off",
    "space-unary-word-ops": "off",
    "spaced-comment": Array [
      1,
      "always",
      Object {
        "block": Object {
          "balanced": true,
          "markers": Array [
            ":",
            "::",
          ],
        },
      },
    ],
    "strict": 1,
    "switch-colon-spacing": "off",
    "symbol-description": 1,
    "template-curly-spacing": "off",
    "template-tag-spacing": "off",
    "unicode-bom": "off",
    "use-isnan": 2,
    "valid-jsdoc": 0,
    "valid-typeof": 0,
    "vars-on-top": 0,
    "wrap-iife": "off",
    "wrap-regex": "off",
    "yield-star-spacing": "off",
    "yoda": 1,
  },
  "settings": Object {
    "import/resolver": Object {
      "node": Object {
        "extensions": Array [
          ".js",
          ".jsx",
          ".android.js",
          ".ios.js",
          ".native.js",
          ".web.js",
        ],
      },
    },
    "react": Object {
      "version": Any<String>,
    },
  },
}
`;
